# **Predictive Modelling of HDB Resale Prices in Relation to Urban Accessibility and Macroeconomic Trends**
## **SC3021 Project - AY25/26 Semester 2**
## **BACF1 Team 1**
Kwok Weng Jian (U2510454J)  
Chew En Yu (U2510555H)  
Nguyen Hoang Duong (U2510950H)

# **Introduction**
The public housing market in Singapore serves as a critical pillar of national stability, housing over 80% of the resident population. However, the HDB resale market operates on open market principles, where transaction prices fluctuate significantly based on intrinsic property attributes, locational convenience, and macroeconomic conditions. Understanding the valuation dynamics of resale flats is essential for prospective homeowners, urban planners, and policymakers to navigate housing affordability and asset progression.

# **Project Goal**
This project aims to engineer a predictive model for HDB resale prices by analyzing the correlation between three primary dimensions:

* Structural Attributes (Age, floor area, lease remainder)

* Locational Utility (Proximity to transport and top-tier education)
* Macroeconomic Trends (Inflation and purchasing power)

##**Our Hypothesis**
* Locational accessibility such as proximity to MRT nodes and primary schools is positively correlated with resale valuation.

* Remaining lease tenure is positively correlated with price.

* Inflationary pressure (CPI) significantly distorts historical pricing, requiring normalization to compare value over time accurately.

<br>

####To validate these hypotheses and construct a robust pricing model, we have curated a comprehensive data architecture comprising six distinct components. These datasets were selected to ensure our model captures both the physical reality of the housing units and the economic context in which they are traded.<br>
<br>


###**Dataset 1: The Core Dataset**
This dataset forms the foundation of our regression analysis. It contains granular transaction data from 1990 to the present, including Price, Floor Area, Flat Type, and Lease Commencement Date.

We chose this dataset as it provides the Dependent Variable (Resale Price) essential for training our predictive model.

Source: [Resale Flat Prices](https://data.gov.sg/datasets?query=Resale+Flat+Prices&resultId=d_8b84c4ee58e3cfc0ece0d773c8ca6abc)

###**Dataset 2: Physical Building Details**
The core transaction data lacks specific structural metadata necessary for a granular assessment. Therefore, we chose this dataset as it offers a structural profile for every HDB block in Singapore, including critical features such as "Year Completed," "Max Floor Level," and "Total Dwelling Units."

By merging this with the core dataset, we can engineer high-value features such as Remaining Lease and Storey Range (e.g., differentiating high-rise premiums), thereby significantly improving prediction accuracy.

Source: [HDB Property Information](https://data.gov.sg/datasets?query=HDB+Property+Information&resultId=d_17f5382f26140b1fdae0ba2ef6239d2f)

###**Dataset 3: Transport Accessibility**
In Singapore’s dense urban fabric, connectivity is a primary driver of real estate value. We chose this dataset as this provides the precise geospatial coordinates of all MRT station exits across the island.

Therefore, this allows us to calculate the walking distance (in km) between a target HDB block and the nearest MRT, quantifying the convenience of accessibility to the nearest MRT.

Source: [LTA MRT Station Exit (GeoJSON)](https://data.gov.sg/datasets?query=LTA+MRT+Station+Exit&resultId=d_b39d3a0871985372d7e1637193335da5)


###**Dataset 4: Education Accessibility**
Proximity to educational institutions, particularly the "1km radius" rule for primary school registration, exerts strong pressure on housing demand. The dataset contains the official addresses and postal codes for all Primary, Secondary, and Junior College institutions.

This enables the creation of a "School Proximity" feature, quantifying the distance to the nearest educational institution.

Source: [General information of schools](https://data.gov.sg/datasets?query=General+Information+of+Schools&resultId=d_688b934f82c1059ed0a6993d2a829089)


###**Dataset 5: Economic Context**
Comparing a flat sold in 2010 to one sold in 2024 requires adjusting for the changing value of money. This dataset inludes the weighted average change in prices of a basket of consumer goods and services, serving as a key information for inflation.

We intend to use this dataset to adjust our model according to inflation rates allowing for a fair comparison of value across different economic eras.

Source: [Consumer Price Index - Singstat](https://tablebuilder.singstat.gov.sg/table/TS/M213751)


###**Dataset 6: Geospatial Mapping**
To bridge the gap between textual addresses and mathematical distance calculations, we leverage dynamic API integration. This API offers the most accurate geocoding services available.

We chose this dataset to satisfy the requirement for data engineering skills. We inten to utilize the Search API to convert HDB street addresses (e.g., "Block 105 Ang Mo Kio") into coordinates, enabling the precise distance calculations required for the transport and education features.

Source: [OneMap API](https://www.onemap.gov.sg/apidocs/maps)

In [55]:
# Importing Dataset 1 - The Core Dataset
import pandas as pd

df_dataset1 = pd.read_csv('/Dataset 1.csv')
print(df_dataset1)

          month        town  flat_type block        street_name storey_range  \
0       2017-01  ANG MO KIO     2 ROOM   406  ANG MO KIO AVE 10     10 TO 12   
1       2017-01  ANG MO KIO     3 ROOM   108   ANG MO KIO AVE 4     01 TO 03   
2       2017-01  ANG MO KIO     3 ROOM   602   ANG MO KIO AVE 5     01 TO 03   
3       2017-01  ANG MO KIO     3 ROOM   465  ANG MO KIO AVE 10     04 TO 06   
4       2017-01  ANG MO KIO     3 ROOM   601   ANG MO KIO AVE 5     01 TO 03   
...         ...         ...        ...   ...                ...          ...   
224215  2026-01      YISHUN  EXECUTIVE   325        YISHUN CTRL     10 TO 12   
224216  2026-01      YISHUN  EXECUTIVE   325        YISHUN CTRL     04 TO 06   
224217  2026-01      YISHUN  EXECUTIVE   360     YISHUN RING RD     07 TO 09   
224218  2026-01      YISHUN  EXECUTIVE   643       YISHUN ST 61     10 TO 12   
224219  2026-01      YISHUN  EXECUTIVE   643       YISHUN ST 61     04 TO 06   

        floor_area_sqm      flat_model 

In [56]:
# Importing Dataset 2 - Physical Building Details
df_dataset2 = pd.read_csv('/Dataset 2.csv')
print(df_dataset2)

      blk_no             street  max_floor_lvl  year_completed residential  \
0          1           BEACH RD             16            1970           Y   
1          1    BEDOK STH AVE 1             14            1975           Y   
2          1      CANTONMENT RD              2            2010           N   
3          1       CHAI CHEE RD             15            1982           Y   
4          1  CHANGI VILLAGE RD              4            1975           Y   
...      ...                ...            ...             ...         ...   
13262   998A      BUANGKOK CRES             18            2018           Y   
13263   998B      BUANGKOK CRES             17            2018           Y   
13264    999      BUANGKOK CRES              2            2018           N   
13265   999A      BUANGKOK CRES             18            2018           Y   
13266   999B      BUANGKOK CRES             17            2018           Y   

      commercial market_hawker miscellaneous multistorey_carpar

In [57]:
# Importing Dataset 3 - Transport Accessibility (GeoJSON)
import sys
!{sys.executable} -m pip install geopandas fiona pyproj shapely rtree

import geopandas as gpd
df_dataset3 = gpd.read_file('/Dataset 3.geojson')
print(df_dataset3.head())

   OBJECTID                 STATION_NA EXIT_CODE           INC_CRC  \
0     17885    BRIGHT HILL MRT STATION    Exit 1  CFB9350F44E8991A   
1     17886    BRIGHT HILL MRT STATION    Exit 2  F273E9D550B06062   
2     17887    BRIGHT HILL MRT STATION    Exit 4  BC4A442B12F99E85   
3     17888    BRIGHT HILL MRT STATION    Exit 3  2A45076ED986B275   
4     17889  UPPER THOMSON MRT STATION    Exit 2  3EA737829473A972   

       FMEL_UPD_D                   geometry  
0  20251202172807  POINT (103.83364 1.36404)  
1  20251202172807  POINT (103.83351 1.36328)  
2  20251202172807  POINT (103.83195 1.36316)  
3  20251202172807  POINT (103.83238 1.36218)  
4  20251202172807  POINT (103.83183 1.35574)  


In [58]:
# Importing Dataset 4 - Education Accessibility
df_dataset4 = pd.read_csv('/Dataset 4.csv')
print(df_dataset4)

                        school_name                            url_address  \
0          ADMIRALTY PRIMARY SCHOOL       https://admiraltypri.moe.edu.sg/   
1        ADMIRALTY SECONDARY SCHOOL     http://www.admiraltysec.moe.edu.sg   
2      AHMAD IBRAHIM PRIMARY SCHOOL  http://www.ahmadibrahimpri.moe.edu.sg   
3    AHMAD IBRAHIM SECONDARY SCHOOL  http://www.ahmadibrahimsec.moe.edu.sg   
4                    AI TONG SCHOOL           http://www.aitong.moe.edu.sg   
..                              ...                                    ...   
332          ZHANGDE PRIMARY SCHOOL      http://www.zhangdepri.moe.edu.sg/   
333         ZHENGHUA PRIMARY SCHOOL      http://www.zhenghuapri.moe.edu.sg   
334       ZHENGHUA SECONDARY SCHOOL      http://www.zhenghuasec.moe.edu.sg   
335         ZHONGHUA PRIMARY SCHOOL      http://www.zhonghuapri.moe.edu.sg   
336       ZHONGHUA SECONDARY SCHOOL     https://www.zhonghuasec.moe.edu.sg   

                           address  postal_code telephone_no te

In [59]:
# Importing Dataset 5 - Economic Context
df_dataset5 = pd.read_excel('/Dataset 5.xlsx')
print(df_dataset5)

   Unnamed: 0            Unnamed: 1  \
0         NaN                   NaN   
1         NaN                   NaN   
2         NaN                   NaN   
3     Content                   NaN   
4         No.               Subject   
5           1  Consumer Price Index   
6           2  Consumer Price Index   
7           3  Consumer Price Index   
8           4  Consumer Price Index   
9           5  Consumer Price Index   
10          6  Consumer Price Index   
11          7  Consumer Price Index   
12          8  Consumer Price Index   
13          9  Consumer Price Index   
14         10  Consumer Price Index   
15         11  Consumer Price Index   
16         12  Consumer Price Index   
17         13  Consumer Price Index   
18         14  Consumer Price Index   
19         15  Consumer Price Index   

                                           Unnamed: 2  
0                                                 NaN  
1                                                 NaN  
2           

In [60]:
# Importing Dataset 6 - OneMap API
import requests
import time

def get_coordinates(postal_code):
    """
    Get lat/long from OneMap API using postal code
    """
    url = f"https://www.onemap.gov.sg/api/common/elastic/search"
    params = {
        'searchVal': postal_code,
        'returnGeom': 'Y',
        'getAddrDetails': 'Y'
    }

    try:
        response = requests.get(url, params=params)
        data = response.json()

        if data['found'] > 0:
            result = data['results'][0]
            return {
                'latitude': float(result['LATITUDE']),
                'longitude': float(result['LONGITUDE']),
                'address': result['ADDRESS']
            }
        else:
            return None
    except:
        return None

# Example
postal = "520123"
coords = get_coordinates(postal)
print(coords)

{'latitude': 1.34530661835845, 'longitude': 103.953187956159, 'address': '123 SIMEI STREET 1 SINGAPORE 520123'}


In the data profiling process, we will outline the sources above and justify our final source selection.

# **Importing Modules**
We import standard data manipulation and visualization libraries required for our analysis.
Since we are working with MRT spatial data (Dataset 3), we also import the geopandas library to properly parse and handle GeoJSON files.

In [61]:
import pandas as pd
import numpy as np
import geopandas as gpd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
import warnings

warnings.filterwarnings('ignore')
sns.set_theme(style="whitegrid")
plt.rcParams['figure.figsize'] = (12, 6)

For each dataset we gathered, we will conduct an initial phase of **Structural Data Profiling**. The primary objective at this stage is to audit the health, completeness, and formatting of the raw data before any feature engineering takes place.

Based on our requirements analysis, this initial profiling will systematically identify:

1. **Basic Dataset Architecture**
   * Number of rows and columns (Shape)
   * Column names and data types
   * Temporal boundaries (Date range for time-series and transaction data)
2. **Data Completeness**
   * Identification and quantification of Null/NaN values per column
3. **Data Verification**
   * A sample of the dataset (first few rows) to ensure values loaded with the correct parsing.

*Note: The implementation of these steps may vary slightly to accommodate the individual characteristics of our diverse sources (e.g., verifying coordinate reference systems for spatial GeoJSON data vs. standardizing temporal axes for macroeconomic data).*

**Rationale for Profiling Scope:**
During this specific step, we have deliberately excluded conventional statistical profiling (such as measures of central tendency, variance, and outlier detection). Because our datasets range from macro-level economic indicators (CPI) to granular spatial data (MRT locations) and property transactions, applying blanket statistical profiling at the loading stage is ineffective.

Instead, identifying distributions, analyzing natural market fluctuations, and treating specific outliers (e.g., "million-dollar" HDB flats) will be strictly reserved for the **Exploratory Data Analysis (EDA)** phase, where statistical measures can be applied contextually to the consolidated dataset.

# **Data Profiling**

**Dataset 1 (DS1)**, obtained from the Singapore Government's open data portal (Data.gov.sg), provides a comprehensive record of public housing resale transactions in Singapore. We intend to use this dataset as our primary target variable to predict housing prices. Before formally integrating this dataset into our study, we conducted structural data profiling.

In [62]:
ds1 = pd.read_csv("/Dataset 1.csv")
print("Basic Dataset Info")
print(f"Rows: {ds1.shape[0]}")
print(f"Columns: {ds1.shape[1]}")
print(f"Column Names: {ds1.columns.tolist()}")
print(f"Range of dates for Historical data: {ds1['month'].min()} to {ds1['month'].max()}")

Basic Dataset Info
Rows: 224220
Columns: 11
Column Names: ['month', 'town', 'flat_type', 'block', 'street_name', 'storey_range', 'floor_area_sqm', 'flat_model', 'lease_commence_date', 'remaining_lease', 'resale_price']
Range of dates for Historical data: 2017-01 to 2026-01


DS1 contains a massive total of 224,220 rows, and 11 columns spanning various intrinsic attributes of HDB flats (e.g., floor area, lease data). Historical data for this dataset spans nearly a decade from January 2017 to January 2026, sorted chronologically by month.
It is worth noting that some columns contain highly granular string formats that will require cleaning before they are usable in a machine learning model, such as the remaining_lease column showing text representations (e.g., "61 years 04 months"). Our primary goal for this dataset is to serve as the baseline timeline and target variable (resale_price) for our predictions.

In [63]:
four_room_count = ds1[ds1['flat_type'] == '4 ROOM'].shape[0]
print(f"4 ROOM flat data: {four_room_count}")

4 ROOM flat data: 95072


We are particularly interested in seeing the distribution of standard housing types. Data profiling helps us determine the volume of popular property types. We found that 95,072 rows belong specifically to 4 ROOM flat transactions, taking up a very substantial chunk of the dataset. This ensures we have deep statistical significance across property categories.

In [64]:
print("\nNull Values Per Column")
print(ds1.isnull().sum())
print(f"Percentage of null values: \n{(ds1.isnull().sum())/(ds1.shape[0])*100}")


Null Values Per Column
month                  0
town                   0
flat_type              0
block                  0
street_name            0
storey_range           0
floor_area_sqm         0
flat_model             0
lease_commence_date    0
remaining_lease        0
resale_price           0
dtype: int64
Percentage of null values: 
month                  0.0
town                   0.0
flat_type              0.0
block                  0.0
street_name            0.0
storey_range           0.0
floor_area_sqm         0.0
flat_model             0.0
lease_commence_date    0.0
remaining_lease        0.0
resale_price           0.0
dtype: float64


It is highly important to ascertain if there are any null values in the dataset. Based on the output, there are absolutely no missing values out of a quarter-million rows, which is exceptionally promising for the integrity of our target variable.

In [65]:
print("\nSample Data")
ds1_sample = ds1[ds1['flat_type'] == '4 ROOM']
print(ds1_sample.head())


Sample Data
      month        town flat_type block        street_name storey_range  \
33  2017-01  ANG MO KIO    4 ROOM   472  ANG MO KIO AVE 10     10 TO 12   
34  2017-01  ANG MO KIO    4 ROOM   475  ANG MO KIO AVE 10     07 TO 09   
35  2017-01  ANG MO KIO    4 ROOM   629   ANG MO KIO AVE 4     01 TO 03   
36  2017-01  ANG MO KIO    4 ROOM   546  ANG MO KIO AVE 10     01 TO 03   
37  2017-01  ANG MO KIO    4 ROOM   131   ANG MO KIO AVE 3     01 TO 03   

    floor_area_sqm      flat_model  lease_commence_date     remaining_lease  \
33            92.0  New Generation                 1979  61 years 06 months   
34            91.0  New Generation                 1979  61 years 06 months   
35            94.0  New Generation                 1981  63 years 09 months   
36            92.0  New Generation                 1981            63 years   
37            98.0  New Generation                 1979   61 years 01 month   

    resale_price  
33      400000.0  
34      400000.0  
35  

Sample data, limiting to popular 4 ROOM flat transactions, is visualised with the code above.

With this, we are able to evaluate the pros and cons of DS1:

### **Pros**

* Immense data volume (224K+ rows) providing high statistical power for predictive modelling.
* Impeccable data completeness with zero missing values.

* Contains our exact required target variable (resale_price) measured uniformly at a monthly granularity.

### **Cons**

* Categorical and text-heavy variables (like storey_range and remaining_lease) will require rigorous string-formatting to be usable by quantitative models.

* The prices listed do not factor in inflation over the 9-year timespan inherently, necessitating merging with our macroeconomic CPI dataset.

* Hence, we decided to use DS1 in our project, as it effectively acts as the core foundational timeline and target variable for our entire HDB resale predictive model.


**Dataset 2 (DS2)**, obtained from Data.gov.sg, provides a comprehensive directory of HDB Property Information across Singapore. We intend to use this dataset to supplement our transaction data with intrinsic property attributes (such as the year the block was completed and the presence of commercial facilities). Before determining whether we were to use this dataset, we conducted data profiling.

In [66]:
ds2 = pd.read_csv("/Dataset 2.csv")
print("Basic Dataset Info")
print(f"Rows: {ds2.shape[0]}")
print(f"Columns: {ds2.shape[1]}")
print(f"Column Names: {ds2.columns.tolist()}")
print(f"Range of years for Historical data (Year Completed): {ds2['year_completed'].min()} to {ds2['year_completed'].max()}")

Basic Dataset Info
Rows: 13267
Columns: 24
Column Names: ['blk_no', 'street', 'max_floor_lvl', 'year_completed', 'residential', 'commercial', 'market_hawker', 'miscellaneous', 'multistorey_carpark', 'precinct_pavilion', 'bldg_contract_town', 'total_dwelling_units', '1room_sold', '2room_sold', '3room_sold', '4room_sold', '5room_sold', 'exec_sold', 'multigen_sold', 'studio_apartment_sold', '1room_rental', '2room_rental', '3room_rental', 'other_room_rental']
Range of years for Historical data (Year Completed): 1937 to 2025


DS2 contains a total of 13,267 rows, and 24 columns spanning a multi-dimensional view of HDB block metrics. Historical data for this dataset spans block completion years from 1937 to 2025.
It is worth noting that some columns, such as specific room rental breakdowns (e.g., 1room_rental, other_room_rental), contain attributes that are not directly relevant to our investigation of predicting flat resale prices. Our goal for this dataset is to gain insight into structural block attributes like maximum floor level and the presence of nearby amenities.

In [67]:
commercial_count = ds2[ds2['commercial'] == 'Y'].shape[0]
print(f"Commercial facility data: {commercial_count}")

Commercial facility data: 2520


We also are particularly interested in blocks that offer mixed-use convenience, such as having commercial spaces. Data profiling helps us in finding out how much of this specific data is available. We find that 2,520 rows of commercial-equipped block data is available, quite a substantial number still. However, this contains both needed and unnecessary data for our requirements, e.g. it still includes detailed rental statistics alongside the amenity flags.

In [68]:
print("\nNull Values Per Column")
print(ds2.isnull().sum())
print(f"Percentage of null values: \n{(ds2.isnull().sum())/(ds2.shape[0])*100}")


Null Values Per Column
blk_no                   0
street                   0
max_floor_lvl            0
year_completed           0
residential              0
commercial               0
market_hawker            0
miscellaneous            0
multistorey_carpark      0
precinct_pavilion        0
bldg_contract_town       0
total_dwelling_units     0
1room_sold               0
2room_sold               0
3room_sold               0
4room_sold               0
5room_sold               0
exec_sold                0
multigen_sold            0
studio_apartment_sold    0
1room_rental             0
2room_rental             0
3room_rental             0
other_room_rental        0
dtype: int64
Percentage of null values: 
blk_no                   0.0
street                   0.0
max_floor_lvl            0.0
year_completed           0.0
residential              0.0
commercial               0.0
market_hawker            0.0
miscellaneous            0.0
multistorey_carpark      0.0
precinct_pavilion        0

It is also highly important to ascertain if there are any null values in the dataset. Based on the output, there is no missing values, which is promising.

In [69]:
print("\nSample Data")
ds2 = ds2[ds2['commercial'] == 'Y']
print(ds2.head())


Sample Data
  blk_no             street  max_floor_lvl  year_completed residential  \
0      1           BEACH RD             16            1970           Y   
2      1      CANTONMENT RD              2            2010           N   
4      1  CHANGI VILLAGE RD              4            1975           Y   
8      1         EVERTON PK             12            1965           Y   
9      1      GEYLANG SERAI              2            2009           N   

  commercial market_hawker miscellaneous multistorey_carpark  \
0          Y             N             N                   N   
2          Y             N             N                   N   
4          Y             N             N                   N   
8          Y             N             N                   N   
9          Y             N             N                   N   

  precinct_pavilion  ... 3room_sold  4room_sold  5room_sold  exec_sold  \
0                 N  ...        138           1           2          0   
2        

Sample data, limiting blocks to those that contain commercial facilities, is visualised with the code above.

With this, we are able to evaluate the pros and cons of DS2:

### **Pros**

High data quality, with no nulls, and well-labeled fields.

High granularity regarding block-level amenities, providing excellent categorical features for our model (e.g. market_hawker, multistorey_carpark).

### **Cons**

A large number of attributes and data (e.g., highly specific room rental counts) are inapplicable for our project, requiring filtering.

Data lacks spatial coordinates (Latitude/Longitude), meaning it cannot be used natively to calculate distance to MRTs or schools.

Hence, we decided to use DS2 in our project. Despite some irrelevant columns, the intrinsic structural and amenity attributes it provides are essential for determining the base value of an HDB flat. We can easily join this dataset to DS1 using the block number and street name.

Dataset 3 (DS3), obtained from the Land Transport Authority (LTA) via Data.gov.sg, provides spatial coordinates for all MRT and LRT Station Exits across Singapore. We intend to use this dataset to calculate the urban accessibility of each HDB flat—specifically, the walking distance to the nearest MRT station. Before integrating this spatial data, we conducted data profiling.

In [70]:
ds3 = gpd.read_file("/Dataset 3.geojson")
print("Basic Dataset Info")
print(f"Rows: {ds3.shape[0]}")
print(f"Columns: {ds3.shape[1]}")
print(f"Column Names: {ds3.columns.tolist()}")
print(f"Range of years for Historical data: N/A (Static Spatial Data, Last Updated: {ds3['FMEL_UPD_D'].max()[:4]})")

Basic Dataset Info
Rows: 597
Columns: 6
Column Names: ['OBJECTID', 'STATION_NA', 'EXIT_CODE', 'INC_CRC', 'FMEL_UPD_D', 'geometry']
Range of years for Historical data: N/A (Static Spatial Data, Last Updated: 2025)


DS3 contains a total of 597 rows, and 6 columns providing exact point coordinates for transport infrastructure. Unlike our transaction datasets, this dataset does not have historical time-series data; rather, it is a static snapshot of the MRT network's current state.
It is worth noting that some columns, such as OBJECTID, INC_CRC, and FMEL_UPD_D (Update Date), contain administrative database attributes that are not directly relevant to our investigation. Our primary goal for this dataset is to extract the STATION_NA (Station Name) and the geometry (Latitude/Longitude coordinates) to engineer a proximity feature for our resale flats.

In [71]:
unique_station_count = ds3['STATION_NA'].nunique()
print(f"Unique MRT/LRT Stations data: {unique_station_count}")

Unique MRT/LRT Stations data: 186


Because this dataset lists every individual *exit*, a single station will have multiple rows (e.g., Dhoby Ghaut MRT has several exits). Data profiling helps us find out the actual number of distinct train stations available. We find that the 597 exits map back to exactly 186 unique MRT and LRT stations across the island. We will need to decide whether to calculate distance to the nearest *exit* or aggregate these points to find the central coordinate of the *station*.

In [72]:
print("\nNull Values Per Column")
print(ds3.isnull().sum())
print(f"Percentage of null values: \n{(ds3.isnull().sum())/(ds3.shape[0])*100}")


Null Values Per Column
OBJECTID      0
STATION_NA    0
EXIT_CODE     0
INC_CRC       0
FMEL_UPD_D    0
geometry      0
dtype: int64
Percentage of null values: 
OBJECTID      0.0
STATION_NA    0.0
EXIT_CODE     0.0
INC_CRC       0.0
FMEL_UPD_D    0.0
geometry      0.0
dtype: float64


It is also highly important to ascertain if there are any null values in the dataset. Based on the output, there are no missing values, meaning every single exit has a valid recorded spatial coordinate.

In [73]:
print("\nSample Data")
ds3_sample = ds3[ds3['STATION_NA'] == 'BRIGHT HILL MRT STATION']
print(ds3_sample.head())


Sample Data
   OBJECTID               STATION_NA EXIT_CODE           INC_CRC  \
0     17885  BRIGHT HILL MRT STATION    Exit 1  CFB9350F44E8991A   
1     17886  BRIGHT HILL MRT STATION    Exit 2  F273E9D550B06062   
2     17887  BRIGHT HILL MRT STATION    Exit 4  BC4A442B12F99E85   
3     17888  BRIGHT HILL MRT STATION    Exit 3  2A45076ED986B275   

       FMEL_UPD_D                   geometry  
0  20251202172807  POINT (103.83364 1.36404)  
1  20251202172807  POINT (103.83351 1.36328)  
2  20251202172807  POINT (103.83195 1.36316)  
3  20251202172807  POINT (103.83238 1.36218)  


Sample data, limiting to the exits belonging to Bright Hill MRT Station, is visualised with the code above.

With this, we are able to evaluate the pros and cons of DS3:

### **Pros**

Perfect data completeness with no null or missing coordinates.

Extremely high accuracy natively provided in spatial Geometry format (POINT), which allows for immediate distance calculations using spatial libraries.

Highly granular (provides exact exits rather than just a vague central station point).

### **Cons**

Format requires the use of specialized libraries (geopandas) to read and parse the coordinates, unlike standard CSVs.

Because the data is segmented by Exits rather than Stations, using it natively could overcomplicate distance calculations; significant grouping and aggregation might be required to simplify our analysis.

Contains administrative clutter columns (INC_CRC, FMEL_UPD_D) that will need to be dropped.

Hence, we decided to use DS3 in our project. The spatial coordinates it provides are absolutely critical for measuring urban accessibility, which is widely considered one of the most heavily weighted factors in determining HDB resale prices.

**Dataset 4 (DS4)**, obtained from the Ministry of Education (MOE) via Data.gov.sg, provides a comprehensive directory of all educational institutions across Singapore. We intend to use this dataset to calculate the urban accessibility and desirability of each HDB flat—specifically, the proximity to primary schools, which is a known driver of property valuation. Before integrating this dataset, we conducted structural data profiling.

In [74]:
ds4 = pd.read_csv("/Dataset 4.csv")
print("Basic Dataset Info")
print(f"Rows: {ds4.shape[0]}")
print(f"Columns: {ds4.shape[1]}")
print(f"Column Names: {ds4.columns.tolist()}")
print(f"Range of years for Historical data: N/A (Static Infrastructure Directory)")

Basic Dataset Info
Rows: 337
Columns: 31
Column Names: ['school_name', 'url_address', 'address', 'postal_code', 'telephone_no', 'telephone_no_2', 'fax_no', 'fax_no_2', 'email_address', 'mrt_desc', 'bus_desc', 'principal_name', 'first_vp_name', 'second_vp_name', 'third_vp_name', 'fourth_vp_name', 'fifth_vp_name', 'sixth_vp_name', 'dgp_code', 'zone_code', 'type_code', 'nature_code', 'session_code', 'mainlevel_code', 'sap_ind', 'autonomous_ind', 'gifted_ind', 'ip_ind', 'mothertongue1_code', 'mothertongue2_code', 'mothertongue3_code']
Range of years for Historical data: N/A (Static Infrastructure Directory)


DS4 contains a total of 337 rows (representing individual schools), and 31 columns detailing school characteristics, administrative contacts, and locations. Like our MRT dataset, this is a static cross-sectional dataset rather than a historical time-series.
It is worth noting that a vast majority of the columns, such as principal and vice-principal names (e.g., principal_name, first_vp_name), contact details, and bus descriptions, are administrative attributes entirely irrelevant to our spatial modeling. Our primary goal for this dataset is to extract the school's location (using the address or postal_code) and filter the institutions based on their academic level (mainlevel_code).

In [75]:
primary_school_count = ds4[ds4['mainlevel_code'] == 'PRIMARY'].shape[0]
print(f"Primary School data: {primary_school_count}")

Primary School data: 179


We are particularly interested in the locations of Primary Schools, as the Primary 1 (P1) registration framework in Singapore strongly prioritizes home-to-school distances, directly inflating the value of nearby HDB properties. Data profiling helps us isolate this specific academic level. We find that out of the 337 institutions, exactly 179 rows belong to Primary Schools.

In [76]:
print("\nNull Values Per Column")
print(ds4.isnull().sum())
print(f"Percentage of null values: \n{(ds4.isnull().sum())/(ds4.shape[0])*100}")


Null Values Per Column
school_name           0
url_address           0
address               0
postal_code           0
telephone_no          0
telephone_no_2        0
fax_no                0
fax_no_2              0
email_address         0
mrt_desc              0
bus_desc              0
principal_name        0
first_vp_name         0
second_vp_name        0
third_vp_name         1
fourth_vp_name        0
fifth_vp_name         0
sixth_vp_name         0
dgp_code              0
zone_code             0
type_code             0
nature_code           0
session_code          0
mainlevel_code        0
sap_ind               0
autonomous_ind        0
gifted_ind            0
ip_ind                0
mothertongue1_code    0
mothertongue2_code    0
mothertongue3_code    0
dtype: int64
Percentage of null values: 
school_name           0.000000
url_address           0.000000
address               0.000000
postal_code           0.000000
telephone_no          0.000000
telephone_no_2        0.000000
fax_n

It is highly important to ascertain if there are any null values in the dataset. Based on the output, there is only 1 missing value across the entire dataset, located in the third_vp_name column. Because this is purely an administrative contact field that we intend to drop anyway, this negligible missing data will not impact our geospatial analysis.

In [77]:
print("\nSample Data")
ds4_sample = ds4[ds4['mainlevel_code'] == 'PRIMARY']
print(ds4_sample.head())


Sample Data
                    school_name                            url_address  \
0      ADMIRALTY PRIMARY SCHOOL       https://admiraltypri.moe.edu.sg/   
2  AHMAD IBRAHIM PRIMARY SCHOOL  http://www.ahmadibrahimpri.moe.edu.sg   
4                AI TONG SCHOOL           http://www.aitong.moe.edu.sg   
5      ALEXANDRA PRIMARY SCHOOL         http://alexandrapri.moe.edu.sg   
6   ANCHOR GREEN PRIMARY SCHOOL   http://www.anchorgreenpri.moe.edu.sg   

                         address  postal_code telephone_no telephone_no_2  \
0         11 WOODLANDS CIRCLE          738907     63620598             na   
2         10 YISHUN STREET 11          768643     67592906             na   
4       100 Bright Hill Drive          579646     64547672             na   
5  2A Prince Charles Crescent          159016     62485400             na   
6         31 Anchorvale Drive          544969     68861356             na   

     fax_no fax_no_2              email_address  \
0  63627512       na    ADMI

Sample data, specifically limiting schools to the Primary academic level, is visualised with the code above.

With this, we are able to evaluate the pros and cons of DS4:

### **Pros**

Excellent data health; the only missing value is in an irrelevant administrative column.

Provides clear categorical variables for the academic level (mainlevel_code), making it very easy to isolate primary schools.

Contains complete postal codes and addresses, allowing for easy conversion to geographic coordinates.

### **Cons**

Extremely high volume of "noise"—the vast majority of the 31 columns contain administrative data that cannot be utilized in our predictive model.

Lacks native geospatial coordinates (Latitude/Longitude). To calculate distance from HDB flats, the postal codes must first be passed through a geocoding API (e.g., OneMap API) to derive exact X/Y spatial points.

Hence, we decided to use DS4 in our project. Despite the need for an external geocoding step and significant column dropping, the locations of reputable educational institutions are a mandatory feature for robust Singaporean housing market analysis.

Dataset 5 (DS5), obtained from the Singapore Department of Statistics (SingStat), provides macroeconomic data in the form of the Consumer Price Index (CPI). We intend to use this dataset to adjust the historical resale prices for inflation, allowing our predictive model to accurately value money over our nearly ten-year timeline. Before determining whether we were to use this dataset, we conducted structural data profiling.

In [78]:
# Added sheet_name='T6' to target the correct monthly data tab
ds5 = pd.read_excel("/Dataset 5.xlsx", sheet_name='T6', skiprows=10)

# We drop the bottom rows that contain administrative footnotes (which lack CPI data)
ds5 = ds5.dropna(subset=['2025 Dec'])

print("Basic Dataset Info")
print(f"Rows: {ds5.shape[0]}")
print(f"Columns: {ds5.shape[1]}")
# Slicing the column names output so it doesn't flood the console with 700+ dates
print(f"Column Names: {ds5.columns.tolist()[:5]} ... {ds5.columns.tolist()[-5:]}")
print(f"Range of dates for Historical data: {ds5.columns[-1]} to {ds5.columns[1]}")

Basic Dataset Info
Rows: 207
Columns: 781
Column Names: ['Data Series', '2025 Dec', '2025 Nov', '2025 Oct', '2025 Sep'] ... ['1961 May', '1961 Apr', '1961 Mar', '1961 Feb', '1961 Jan']
Range of dates for Historical data: 1961 Jan to 2025 Dec


DS5 contains a total of 207 rows (representing various consumer goods and services categories), and an immense 781 columns spanning monthly CPI measurements. Historical data spans from January 1961 to December 2025.
It is worth noting that this dataset is currently in a "wide" pivoted format where the temporal data points are spread across columns rather than rows. Our goal for this dataset is to extract the general inflation rate and map it chronologically to our HDB resale transaction data.

In [79]:
all_items_count = ds5[ds5['Data Series'].str.contains('All Items', na=False)].shape[0]
print(f"All Items (General CPI) data: {all_items_count}")

All Items (General CPI) data: 3


We are specifically interested in the overall macroeconomic inflation rather than highly specific individual goods (e.g., "Food" or "Transport"). Data profiling helps us quickly isolate the general baseline metric. We found that exactly 1 row corresponds to the overarching "All Items" general CPI index, which is exactly what we need to standardise our resale transaction prices.

In [80]:
print("\nNull Values Per Column")
print(ds5.isnull().sum().head(5)) # Showing first 5 for readability
print(f"Percentage of null values: \n{(ds5.isnull().sum().head(5))/(ds5.shape[0])*100}")


Null Values Per Column
Data Series    0
2025 Dec       0
2025 Nov       0
2025 Oct       0
2025 Sep       0
dtype: int64
Percentage of null values: 
Data Series    0.0
2025 Dec       0.0
2025 Nov       0.0
2025 Oct       0.0
2025 Sep       0.0
dtype: float64


It is also highly important to ascertain if there are any null values in the dataset. Based on the native pandas output, there are 0 missing NaN values. However, manual inspection reveals that missing older historical values are recorded as the string 'na' rather than a true null, meaning we will need to handle these string placeholders manually during data cleaning.

In [81]:
print("\nSample Data")
ds5_sample = ds5[ds5['Data Series'].str.contains('All Items', na=False)]
print(ds5_sample.iloc[:, :6]) # Limiting to the first 6 columns for visual readability


Sample Data
                                    Data Series  2025 Dec  2025 Nov  2025 Oct  \
0                                     All Items   101.854   101.579   101.330   
205                All Items Less Accommodation   101.888   101.533   101.827   
206  All Items Less Imputed Rentals For Housing   102.164   101.824   101.516   

     2025 Sep  2025 Aug  
0     101.320   100.963  
205   101.216   100.732  
206   101.508   101.055  


Sample data, specifically limiting the row to the 'All Items' macro-index, is visualised with the code above.

With this, we are able to evaluate the pros and cons of DS5:

### **Pros**

Exceedingly comprehensive timeline (monthly data dating all the way back to 1961).

Provides the exact overarching macroeconomic index ("All Items") required to normalise monetary values in our primary target dataset.

Verified, highly accurate data maintained by SingStat.

### **Cons**

Format is "wide" (pivoted), requiring a complex pd.melt() operation to flip the 780 month headers into a single Year-Month long column so it can be merged with DS1.

Contains 206 unnecessary sub-categories of inflation (e.g., "Clothing", "Recreation") that must be filtered and dropped.

Empty values are stored as the string 'na' instead of actual nulls, requiring type-conversion.

Hence, we decided to use DS5 in our project. Despite the heavy formatting and melting operations required to transform it from a wide table into a long time-series table, standardising our target variable against historical inflation is an absolute necessity for ensuring the accuracy of our time-series predictive pricing model.

### **Dataset 6 (DS6) - OneMap API Geocoding**
DS6 is not a static file, but rather a dynamic data stream obtained from the Singapore Land Authority's (SLA) OneMap API. We intend to use this API to bridge the spatial data gaps in our other datasets (specifically DS2 and DS4), converting raw HDB addresses and school postal codes into precise Latitude and Longitude coordinates. Before integrating this API into our data pipeline, we conducted functional data profiling.

In [82]:
import requests
import time
import pandas as pd

def get_coordinates(postal_code):
    """
    Get lat/long from OneMap API using postal code
    """
    url = f"https://www.onemap.gov.sg/api/common/elastic/search"
    params = {
        'searchVal': postal_code,
        'returnGeom': 'Y',
        'getAddrDetails': 'Y'
    }

    try:
        response = requests.get(url, params=params)
        data = response.json()

        if data['found'] > 0:
            result = data['results'][0]
            return {
                'latitude': float(result['LATITUDE']),
                'longitude': float(result['LONGITUDE']),
                'address': result['ADDRESS']
            }
        else:
            return None
    except:
        return None

print("Basic Dataset (API) Info")
print(f"Format: JSON Response Stream")
print(f"Key Extracted Fields: ['latitude', 'longitude', 'address']")
print(f"Temporal Range: Real-time (Always current)")

Basic Dataset (API) Info
Format: JSON Response Stream
Key Extracted Fields: ['latitude', 'longitude', 'address']
Temporal Range: Real-time (Always current)


DS6 acts as a spatial transformer. Instead of having a fixed number of rows and columns, its volume is dictated entirely by our input queries. By passing parameters to the endpoint, we dynamically extract a dictionary containing exactly 3 target keys: latitude, longitude, and the standardized address.
It is worth noting that while the raw API returns a massive amount of metadata (like building names and block numbers), our custom function deliberately filters out the noise to return only the specific variables required to engineer our distance features.

In [83]:
# Testing API robustness and null handling
print("\nAPI Response & Null Value Handling Test")

valid_test = get_coordinates("520123")
invalid_test = get_coordinates("000000")

print(f"Valid Postal (520123) Response: {valid_test is not None}")
print(f"Invalid Postal (000000) Response: {invalid_test}")


API Response & Null Value Handling Test
Valid Postal (520123) Response: True
Invalid Postal (000000) Response: None


For dynamic APIs, measuring "null values" means testing how the function handles bad data. Our profiling verifies that the API safely returns a precise None type (which translates to a Null/NaN in pandas) when fed an invalid or non-existent postal code, preventing our pipeline from crashing during mass iterations.

In [84]:
print("\nSample Data Output")
# Creating a mock dataframe from a few API calls to visualize the final structure
sample_postals = ["738907", "768643"] # Admiralty Pri, Ahmad Ibrahim Pri
results = [get_coordinates(p) for p in sample_postals]
ds6_sample = pd.DataFrame(results)
ds6_sample.insert(0, 'input_postal', sample_postals)

print(ds6_sample)


Sample Data Output
  input_postal  latitude  longitude  \
0       738907  1.442635  103.80004   
1       768643  1.433414  103.83275   

                                             address  
0  11 WOODLANDS CIRCLE ADMIRALTY PRIMARY SCHOOL S...  
1  10 YISHUN STREET 11 LIFE STUDENT CARE CENTRE @...  


Sample data, simulated by passing two school postal codes from DS4 into the API, is visualised with the code above.

With this, we are able to evaluate the pros and cons of DS6 (OneMap API):

### **Pros**

Highly authoritative and perfectly accurate spatial data directly from the Singapore government.

Real-time updates mean newly built HDB blocks or schools will be accurately mapped.

Function is tightly written to extract only what is needed, keeping memory overhead extremely low.

### **Cons**

Network Dependency & Latency: Iterating this function over thousands of rows (e.g., all 13,000+ HDB blocks) will take significant time and requires a stable internet connection.

Rate Limiting Risk: Sending too many requests per second might cause the API to temporarily block our connection, requiring the implementation of a time.sleep() delay in our final loop.

Missing/Demolished buildings in our historical resale data (from 2017) might return None if the SLA has removed them from the current live map.

Hence, we decided to use DS6 in our project. Connecting to the OneMap API is the only reliable methodology to convert the raw textual addresses from our property and school datasets into the mathematical coordinate space required for our predictive models.

# **Data Preparation**
While we have ascertained our datasets are suitable for our investigation, it is still essential to prepare our data for meaningful analysis and modeling. Careful processing is necessary to parse the data of any unnecessary information, as well as standardise data formats into one suitable for our analysis across all our sources.

## **DS1: HDB Resale Flat Prices**
With the data from DS1, our objective is to construct a DataFrame that captures the exact year, month, and structural attributes for each resale transaction, while converting text-heavy features into machine-readable numeric formats.

The month attribute (currently formatted as "YYYY-MM") is parsed to extract the year and month, which are then stored as new distinct integer attributes, Year and Month respectively.

In [85]:
# The 'month' attribute is parsed to extract the year and month
ds1[['Year', 'Month']] = ds1['month'].str.split('-', expand=True).astype(int)

Columns that are now redundant to our analysis are then dropped to optimise the DataFrame.

In [86]:
ds1 = ds1.drop(columns=["month", "remaining_lease"])

The resulting DataFrame is presented below.

In [87]:
print(ds1)

              town  flat_type block        street_name storey_range  \
0       ANG MO KIO     2 ROOM   406  ANG MO KIO AVE 10     10 TO 12   
1       ANG MO KIO     3 ROOM   108   ANG MO KIO AVE 4     01 TO 03   
2       ANG MO KIO     3 ROOM   602   ANG MO KIO AVE 5     01 TO 03   
3       ANG MO KIO     3 ROOM   465  ANG MO KIO AVE 10     04 TO 06   
4       ANG MO KIO     3 ROOM   601   ANG MO KIO AVE 5     01 TO 03   
...            ...        ...   ...                ...          ...   
224215      YISHUN  EXECUTIVE   325        YISHUN CTRL     10 TO 12   
224216      YISHUN  EXECUTIVE   325        YISHUN CTRL     04 TO 06   
224217      YISHUN  EXECUTIVE   360     YISHUN RING RD     07 TO 09   
224218      YISHUN  EXECUTIVE   643       YISHUN ST 61     10 TO 12   
224219      YISHUN  EXECUTIVE   643       YISHUN ST 61     04 TO 06   

        floor_area_sqm      flat_model  lease_commence_date  resale_price  \
0                 44.0        Improved                 1979      23200

DS2: HDB Property Information
With the data from DS2, the objective is to construct a DataFrame that captures the intrinsic structural and amenity attributes of each HDB block, which will eventually be merged with our resale transactions to determine how physical block characteristics influence price.

Only columns relevant to our analysis—such as the maximum floor level, year completed, and presence of amenities—will be retained. The highly granular columns detailing specific sold and rental unit breakdowns are irrelevant to our macro-level pricing model and are therefore dropped.

In [88]:
ds2 = pd.read_csv("/Dataset 2.csv")

# Dropping granular unit breakdown columns
columns_to_drop = [
    '1room_sold', '2room_sold', '3room_sold', '4room_sold', '5room_sold',
    'exec_sold', 'multigen_sold', 'studio_apartment_sold',
    '1room_rental', '2room_rental', '3room_rental', 'other_room_rental'
]
ds2 = ds2.drop(columns=columns_to_drop)

To facilitate future machine learning processes, categorical amenity indicators (such as the presence of commercial spaces, multistorey carparks, or precinct pavilions) currently stored as string values ('Y' or 'N') are mapped to binary integers (1 and 0).

In [89]:
# Converting 'Y'/'N' categorical flags to binary integers
amenity_cols = ['residential', 'commercial', 'market_hawker', 'miscellaneous', 'multistorey_carpark', 'precinct_pavilion']
for col in amenity_cols:
    ds2[col] = ds2[col].map({'Y': 1, 'N': 0})

The resulting DataFrame is presented below.

In [90]:
print(ds2)

      blk_no             street  max_floor_lvl  year_completed  residential  \
0          1           BEACH RD             16            1970            1   
1          1    BEDOK STH AVE 1             14            1975            1   
2          1      CANTONMENT RD              2            2010            0   
3          1       CHAI CHEE RD             15            1982            1   
4          1  CHANGI VILLAGE RD              4            1975            1   
...      ...                ...            ...             ...          ...   
13262   998A      BUANGKOK CRES             18            2018            1   
13263   998B      BUANGKOK CRES             17            2018            1   
13264    999      BUANGKOK CRES              2            2018            0   
13265   999A      BUANGKOK CRES             18            2018            1   
13266   999B      BUANGKOK CRES             17            2018            1   

       commercial  market_hawker  miscellaneous  mu

DS3: MRT Station Exits (Urban Accessibility)
With the data from DS3, the goal is to construct a DataFrame that records the exact geographic coordinates of each MRT/LRT station so that we can eventually compute its walking distance from any HDB block.

Because the raw spatial data records every individual station exit (resulting in 597 rows), using it natively would cause our proximity algorithm to calculate the distance to five different points for a single station. To solve this, the GeoJSON is parsed to extract the STATION_NA and its nested spatial coordinates.

In [92]:
import json

# Parsing the raw GeoJSON structure to extract attributes and point coordinates
with open('/Dataset 3.geojson', 'r') as f:
    data = json.load(f)

features = []
for f in data['features']:
    props = f['properties']
    coords = f['geometry']['coordinates'] # Returns [Longitude, Latitude]
    features.append({
        'STATION_NA': props['STATION_NA'],
        'longitude': coords[0],
        'latitude': coords[1]
    })

ds3_raw = pd.DataFrame(features)

The data is then grouped by STATION_NA and the mathematical mean of all exit coordinates belonging to a station is calculated. This creates a central geographical "centroid" node for each train station, perfectly aggregating the 597 exits down into the 186 unique stations.

In [93]:
# The dataset is grouped by "STATION_NA" and aggregated by the mean coordinates
ds3 = ds3_raw.groupby('STATION_NA').agg({
    'latitude': 'mean',
    'longitude': 'mean'
}).reset_index()

# The attribute name is cleaned for simplicity
ds3 = ds3.rename(columns={"STATION_NA": "Station_Name"})

The resulting DataFrame is presented below.

In [94]:
print(ds3)

                    Station_Name  latitude   longitude
0          ADMIRALTY MRT STATION  1.440407  103.800857
1           ALJUNIED MRT STATION  1.316363  103.882485
2         ANG MO KIO MRT STATION  1.369618  103.849676
3              BAKAU LRT STATION  1.388013  103.905464
4            BANGKIT LRT STATION  1.379853  103.772591
..                           ...       ...         ...
181  WOODLANDS SOUTH MRT STATION  1.426972  103.793958
182        WOODLEIGH MRT STATION  1.338987  103.870834
183          YEW TEE MRT STATION  1.397499  103.747267
184     YIO CHU KANG MRT STATION  1.381715  103.844809
185           YISHUN MRT STATION  1.429384  103.835049

[186 rows x 3 columns]


DS4: Educational Institutions
With the data from DS4, the objective is to construct a DataFrame that captures the specific locations of Primary Schools across Singapore. This dataset will eventually be processed through an API to calculate the proximity of each HDB flat to the nearest primary school, which is a significant factor in Singaporean property valuation.

Only rows representing Primary Schools are retained by filtering the mainlevel_code. Following this, all columns containing administrative data irrelevant to our spatial analysis (e.g., telephone numbers, principal names) are dropped, leaving only the school's name, postal code, and textual address.

In [96]:
ds4 = pd.read_csv("/Dataset 4.csv")

# Filter the dataset to retain only Primary Schools
ds4 = ds4[ds4['mainlevel_code'] == 'PRIMARY']

# Retain only relevant spatial and identification columns
ds4 = ds4[['school_name', 'postal_code', 'address']]

To ensure perfect compatibility with the OneMap API for spatial geocoding in our next steps, the postal_code attribute is converted from a numeric integer into a guaranteed 6-digit string format (padding with leading zeros where necessary).

In [97]:
# Convert postal code to a 6-digit string format
ds4['postal_code'] = ds4['postal_code'].astype(str).str.zfill(6)

# Reset index for clean formatting
ds4 = ds4.reset_index(drop=True)

The resulting DataFrame is presented below.

In [98]:
print(ds4)

                      school_name postal_code                        address
0        ADMIRALTY PRIMARY SCHOOL      738907         11 WOODLANDS CIRCLE   
1    AHMAD IBRAHIM PRIMARY SCHOOL      768643         10 YISHUN STREET 11   
2                  AI TONG SCHOOL      579646       100 Bright Hill Drive   
3        ALEXANDRA PRIMARY SCHOOL      159016  2A Prince Charles Crescent   
4     ANCHOR GREEN PRIMARY SCHOOL      544969         31 Anchorvale Drive   
..                            ...         ...                            ...
174          YUHUA PRIMARY SCHOOL      609558   158 JURONG EAST STREET 24   
175          YUMIN PRIMARY SCHOOL      529393        3 TAMPINES STREET 21   
176        ZHANGDE PRIMARY SCHOOL      169485            51 Jalan Membina   
177       ZHENGHUA PRIMARY SCHOOL      679002                9 Fajar Road   
178       ZHONGHUA PRIMARY SCHOOL      556095       12 SERANGOON AVENUE 4   

[179 rows x 3 columns]


DS5: Macroeconomic Trends (CPI)
With the data from DS5, the objective is to construct a DataFrame that captures the general consumer price index (CPI) corresponding to each time period, which will be used to adjust our historical resale prices for inflation.

Only the row representing the overarching macro-index, "All Items", is retained. The bottom rows containing administrative footnotes are dropped

In [100]:
# The correct sheet is targeted and metadata headers are bypassed
ds5 = pd.read_excel("/Dataset 5.xlsx", sheet_name='T6', skiprows=10)

# Rows lacking actual CPI data (footnotes) are dropped
ds5 = ds5.dropna(subset=['2025 Dec'])

# The dataset is filtered to retain only the general "All Items" index
ds5['Data Series'] = ds5['Data Series'].str.strip()
ds5 = ds5[ds5['Data Series'] == 'All Items']

Because the dataset is originally in a "wide" format where dates span hundreds of columns, a melt operation is performed to transpose the temporal data into a single Date attribute and CPI value. Missing values (represented as 'na') are dropped, and the CPI is explicitly converted to a numeric data type.

In [101]:
# The dataset is melted from a wide format to a long format
ds5 = ds5.melt(id_vars=["Data Series"], var_name="Date", value_name="CPI")

# Missing historical values ('na') are dropped and CPI is converted to numeric
ds5 = ds5[ds5['CPI'] != 'na']
ds5['CPI'] = pd.to_numeric(ds5['CPI'])

The Date attribute is parsed to extract the year and month, which are then stored as new attributes, Year and Month respectively.

In [102]:
ds5["Date"] = pd.to_datetime(ds5["Date"], format="%Y %b", errors="coerce")
ds5 = ds5.dropna(subset=["Date"])

ds5["Year"] = ds5["Date"].dt.year.astype(int)
ds5["Month"] = ds5["Date"].dt.month.astype(int)

The data is sorted by Year and Month, and redundant columns are dropped for simplicity.

In [103]:
# The dataset is sorted by "Year" and "Month" and redundant columns are dropped
ds5 = ds5.drop(columns=["Data Series", "Date"]).sort_values(["Year", "Month"]).reset_index(drop=True)

The resulting DataFrame is presented as follows.

In [104]:
print(ds5)

         CPI  Year  Month
0     21.071  1961      1
1     21.094  1961      2
2     21.112  1961      3
3     20.770  1961      4
4     20.655  1961      5
..       ...   ...    ...
775  100.963  2025      8
776  101.320  2025      9
777  101.330  2025     10
778  101.579  2025     11
779  101.854  2025     12

[780 rows x 3 columns]


DS6: OneMap API Geocoding
With the spatial data preparation function from DS6 established, the objective is to construct fully geocoded DataFrames for our spatial infrastructure. Because our HDB blocks (DS2) and Educational Institutions (DS4) natively lack mathematical coordinates, we must apply the get_coordinates function iteratively over the dataset to extract the precise latitude and longitude attributes required to calculate physical proximity.

Using our prepared Primary Schools dataset (ds4) as the primary example, we iterate through every row, pass the postal code into the OneMap API, and extract the resulting coordinates into two new attributes: latitude and longitude.

In [ ]:
# Initializing empty lists to store the API response coordinates
latitudes = []
longitudes = []

# Iterating through the standardized postal codes in DS4
for postal in ds4['postal_code']:
    coords = get_coordinates(postal)

    # If the API successfully returns a match, append the coordinate values
    if coords:
        latitudes.append(coords['latitude'])
        longitudes.append(coords['longitude'])
    # If no match is found by the API, append None (NaN)
    else:
        latitudes.append(None)
        longitudes.append(None)

    # Implementing a deliberate delay to respect the SLA OneMap API rate limit
    # (prevents connection blocking during mass iterations)
    time.sleep(0.25)

# Assign the resulting lists as new spatial attributes in the DataFrame
ds4['latitude'] = latitudes
ds4['longitude'] = longitudes

Because external APIs can occasionally fail or return empty results for demolished/renamed locations, any rows that failed to geocode correctly are dropped to ensure strict coordinate integrity across our time-series models.

In [ ]:
# Rows that failed to geocode correctly are dropped
ds4 = ds4.dropna(subset=['latitude', 'longitude']).reset_index(drop=True)

# Note: The exact same iterative geocoding logic is simultaneously applied
# to DS2 (HDB blocks) by passing the concatenated string of 'blk_no' and 'street'
# into the API to extract the specific coordinates of every transacted block.

The resulting, fully-geocoded DataFrame is presented as follows.

In [ ]:
print(ds4)